### ISSUE TITLE
serverBareModulesPlugin displays incorrect warnings for subpath imports in 1.6.2

### ISSUE THREAD

**Issue by jacobwgillespie** [remix-run/remix#3604]
### What version of Remix are you using?

1.6.2

### Steps to Reproduce

The change introduced in #3085, where dependencies are tested with `require.resolve`, fails for modules that do not export a default `main`. For instance, my Remix app imports `@noble/hashes/blake3` - the serverBareModulesPlugin will attempt to `require.resolve('@noble/hashes')`, which fails because `@noble/hashes` does not have a main export (intentionally so).

1. Attempt to import `@noble/hahes/blake3` in a file (or any other module that requires importing from a subpath export and does not have a main export)

### Expected Behavior

I would expect that the detection logic is able to determine that `@noble/hashes` is included in my `package.json`

### Actual Behavior

Currently it assumes that any failure of `require.resolve` implies that the module is not listed in `package.json` ([source](https://github.com/remix-run/remix/blob/914792546ff6551b45ac93e1d386087aa6ae1a87/packages/remix-dev/compiler/plugins/serverBareModulesPlugin.ts#L73-L83)) and displays the following warning:

```
The path "@noble/hashes/blake3" is imported in app/utils/filename.server.ts but @noble/hashes is not listed in your package.json dependencies. Did you forget to install it?
```

**Comment by garth** [comment]
Same issue for `@heroicons/react`.

```
The path "@heroicons/react/solid" is imported in src/app/routes/login.tsx but @heroicons/react is not listed in your package.json dependencies. Did you forget to install it?
```

**Comment by cayter** [comment]
Did some initial checks. The current code `require.resolve` would need the library package.json to be like the below:

```json
"main": "...", // this must be configured

OR

"exports": {
  ".": "...", // if the library is using `exports`, then this line must be configured 
}
```

How did this happen? Welcome to NodeJS, we move fast and break everything. The ESM push is half-baked and leave a lot of libraries not fully compatible with each other.

**Comment by machour** [comment]
Tried the nightly, and confirming that this is indeed fixed by #3656, thank you @mcansh !
